api/CMakeLists.txt

```diff
@@ -1,161 +1,232 @@
+    bag_uint8array.h
+    bag_valuetable.h
+    bag_version.h
+)
+source_group("Header Files" FILES ${BAG_HEADER_FILES} ${BAG_PRIVATE_HEADER_FILES})
+
+# Allow for inserting BAG as a nested library (not tested)
+if(BAG_BUILD_SHARED_LIBS)
+    add_library(baglib SHARED
+        ${BAG_SOURCE_FILES} ${BAG_HEADER_FILES} ${BAG_PRIVATE_HEADER_FILES}
+    )
+else()
+    add_library(baglib STATIC
+        ${BAG_SOURCE_FILES} ${BAG_HEADER_FILES} ${BAG_PRIVATE_HEADER_FILES}
+    )
 endif()
 
-#Use the version 1 interface for HDF5
-add_definitions(-D H5_USE_16_API)
-
-IF(MSVC)
-
-    #ignore some warnings generated by windows.
-    add_definitions(/D_CRT_SECURE_NO_WARNINGS /D _CRT_NONSTDC_NO_DEPRECATE)
-    
-	#if we are building a shared library, then we need to define these symbols.
-	IF(BUILD_SHARED_LIBS)
-		add_definitions(/D BAG_DLL /D BAG_EXPORTS)
-	ENDIF(BUILD_SHARED_LIBS)
-	
-	#if we are building against a shared HDF5, then we need to define this symbol.
-	IF(BUILD_WITH_SHARED_HDF5)
-		add_definitions(/D _HDF5USEDLL_)
-	ENDIF(BUILD_WITH_SHARED_HDF5)
-	
-ENDIF(MSVC)
-
-#List all of the bag api source files.
-set (BAG_SOURCE_FILES 
- bag_attr.c
- bag_crypto.c
- bag_hdf.c
- bag_metadata.cpp
- bag_metadata_export.cpp
- bag_metadata_import.cpp
- bag_legacy.c
- bag_opt_group.c
- bag_opt_surfaces.c
- bag_reference_system.cpp
- bag_surface_correct.c
- bag_surfaces.c
- bag_tracking_list.c
- crc32.c
- onscrypto.c)
-source_group("Source Files" FILES ${BAG_SOURCE_FILES})
- 
-set (BAG_HEADER_FILES 
- bag.h
- bag_config.h
- bag_errors.h
- bag_metadata.h
- bag_legacy.h
- bag_opt_surfaces.h
- bag_private.h
- crc32.h
- dt_cc.h
- onscrypto.h
- stdtypes.h)
-source_group("Header Files" FILES ${BAG_HEADER_FILES})
- 
-add_library (bag ${LIB_TYPE} ${BAG_SOURCE_FILES} ${BAG_HEADER_FILES} )
-
-target_link_libraries(bag ${HDF5_LIB})
-target_link_libraries(bag ${BEECRYPT_LIB})
-target_link_libraries(bag ${LIBXML_LIB})
-
-IF(MSVC)
-
-    #Apparently some versions of Visual Studio require us to link to ws2_32.lib
-    target_link_libraries(bag "ws2_32.lib")
-    
-ENDIF(MSVC)
-
-install(FILES ${BAG_HEADER_FILES} DESTINATION include)
-
-install(TARGETS bag
-        RUNTIME DESTINATION bin
-        LIBRARY DESTINATION lib
-        ARCHIVE DESTINATION lib/static)
-        
-install(DIRECTORY ${PROJECT_SOURCE_DIR}/configdata/ DESTINATION configdata)
+set_target_properties(baglib
+    PROPERTIES
+        CXX_STANDARD 14
+        CMAKE_CXX_STANDARD_REQUIRED ON
+        CMAKE_CXX_EXTENSIONS OFF
+)
+
+target_compile_definitions(baglib
+    PRIVATE
+        $<$<CXX_COMPILER_ID:MSVC>:_USE_MATH_DEFINES>
+        $<$<CXX_COMPILER_ID:MSVC>:_CRT_SECURE_NO_WARNINGS>
+        $<$<CXX_COMPILER_ID:MSVC>:_CRT_NONSTDC_NO_DEPRECATE>
+        $<$<AND:$<CXX_COMPILER_ID:MSVC>,$<BOOL:BAG_BUILD_SHARED_LIBS>>:BAG_DLL>
+        $<$<AND:$<CXX_COMPILER_ID:MSVC>,$<BOOL:BAG_BUILD_SHARED_LIBS>>:BAG_EXPORTS>
+        $<$<AND:$<CXX_COMPILER_ID:MSVC>,$<BOOL:BAG_BUILD_SHARED_LIBS>>:H5_BUILT_AS_DYNAMIC_LIB>
```

Comment:
Is here the right place to force this HDF5 flag? Should it be somehow documented with a warning to avoid headaches in library users?

Reply:
We can document that HDF5 must be built as a DLL on Windows for now and someone can re-architect and submit a PR in the future.

Reply:
Resolved: https://github.com/OpenNavigationSurface/BAG/commit/21beafbd5b480acdb16cc01751ca5b99fb95e9c8